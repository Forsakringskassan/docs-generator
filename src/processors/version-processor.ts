import { type Processor } from "../processor";
import { interpolate, runCommand } from "../utils";

/**
 * Options for {@link versionProcessor}.
 *
 * @public
 */
export interface VersionProcessorOptions {
    readonly enabled: boolean;
    readonly scm?: {
        readonly commitUrlFormat: string;
        readonly prUrlFormat: string;
    };
}

/**
 * @internal
 */
export interface VersionProcessorTemplateData {
    readonly pkg: {
        readonly name: string;
        readonly version: string;
    };
    readonly scm: {
        readonly branch: string;
        readonly commitUrl: string;
        readonly commitShort: string;
        readonly commitHash: string;
        readonly pr: string | undefined;
        readonly prUrl: string | undefined;
    } | null;
    readonly build: {
        date: string;
        time: string;
    };
}

type ScmOptions = NonNullable<VersionProcessorOptions["scm"]>;

interface GitData {
    readonly branch: string;
    readonly commitUrl: string;
    readonly commitShort: string;
    readonly commitHash: string;
    readonly pr: string | undefined;
    readonly prUrl: string | undefined;
}

async function getGitBranch(): Promise<string> {
    /* jenkins doesn't checkout using branch name so cant fetch using normal
     * method but instead puts the branch name in environmental variables */
    const { CHANGE_BRANCH, CHANGE_NAME } = process.env;
    if (CHANGE_BRANCH) {
        return CHANGE_BRANCH;
    }
    if (CHANGE_NAME) {
        return CHANGE_NAME;
    }
    const branch = await runCommand(
        "git rev-parse --abbrev-ref HEAD",
        "unknown",
    );
    if (branch === "HEAD" || branch === "") {
        return "unknown";
    } else {
        return branch;
    }
}

function getPullRequestID(): string | undefined {
    /* jenkins */
    if (process.env.JOB_BASE_NAME) {
        const name = process.env.JOB_BASE_NAME;
        const match = name.match(/PR-(\d+)/);
        return match ? match[1] : undefined;
    }

    return undefined;
}

async function getSCMData(
    pkg: { homepage: string },
    options: ScmOptions,
): Promise<GitData | null> {
    const commitHash = await runCommand(`git rev-parse HEAD`);
    const commitShort = await runCommand(`git rev-parse --short HEAD`);
    if (!commitHash || !commitShort) {
        return null;
    }

    const { homepage } = pkg;
    const pr = getPullRequestID();
    const prUrl = pr
        ? interpolate(options.prUrlFormat, { homepage, pr })
        : undefined;
    return {
        branch: await getGitBranch(),
        commitShort,
        commitHash,
        commitUrl: interpolate(options.commitUrlFormat, {
            hash: commitHash,
            short: commitShort,
            homepage,
        }),
        pr,
        prUrl,
    };
}

function getBuildDatetime(): { date: string; time: string } {
    const date = new Date();
    return {
        date: date.toLocaleDateString("sv-SE"),
        time: date.toLocaleTimeString("sv-SE", {
            hour: "2-digit",
            minute: "2-digit",
        }),
    };
}

/**
 * Injects the package name and version into the document.
 *
 * @public
 */
export function versionProcessor(
    pkg: { name: string; version: string; homepage: string },
    container: string,
    options?: VersionProcessorOptions,
): Processor {
    const { enabled = true, scm = false } = options ?? {};
    return {
        name: "version-processor",
        after: "generate-docs",
        async handler(context) {
            if (!enabled) {
                return;
            }
            const scmData = scm ? await getSCMData(pkg, scm) : null;
            const buildData = getBuildDatetime();
            const data: VersionProcessorTemplateData = {
                pkg,
                scm: scmData,
                build: buildData,
            };
            context.addTemplateBlock(container, "version", {
                filename: "partials/version.njk.html",
                data,
            });
        },
    };
}
