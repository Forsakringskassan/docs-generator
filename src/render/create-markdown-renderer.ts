import markdownIt from "markdown-it";
import markdownItDeflist from "markdown-it-deflist";
import { type Document, type DocumentPage } from "../document";
import { type ExampleResult } from "../examples/example-result";
import { processInlineTags } from "./process-inline-tags";
import inlineTags from "./inline-tags";
import {
    codeInline,
    codePreview,
    headingLevel,
    imageResources,
    containerRenderer,
    table,
} from "./markdown";
import { MarkdownEnv } from "./markdown-env";
import { SoftErrorType } from "./soft-error";

/**
 * Options for [[createMarkdownRenderer]].
 *
 * @public
 */
export interface MarkdownOptions {
    /** List of all available documents */
    readonly docs: Document[];

    /** Callback for adding a static resource */
    addResource(dst: string, src: string): void;

    /** Callback when an example should be generated */
    generateExample(options: {
        source: string;
        language: string;
        filename: string;
        tags: string[];
    }): ExampleResult;

    /**
     * Handle error which can be recovered from.
     *
     * If a string is returned it replaces the intended content but otherwise
     * continues rendering.
     *
     * If the exception is rethrown it will cause the renderer to bail out.
     *
     * @param error - The error to be handled.
     * @returns A replacement string or rethrows error.
     */
    handleSoftError(error: SoftErrorType): string;

    /**
     * Options for messagebox container.
     */
    messagebox?: {
        /** Default titles for messageboxes */
        title?: Record<string, string>;
    };
}

/**
 * Markdown renderer.
 *
 * @public
 */
export interface MarkdownRenderer {
    /**
     * Render given markdown content to HTML
     *
     * @param doc - Document metadata.
     * @param content - Markdown content to render.
     * @returns HTML rendered from markdown content.
     */
    render(doc: DocumentPage, content: string): string;
}

/**
 * @public
 */
export function createMarkdownRenderer(
    options: MarkdownOptions,
): MarkdownRenderer {
    const { docs } = options;
    const env: MarkdownEnv = {
        fileInfo: {
            fullPath: "",
            name: "",
            path: "",
            outputName: false,
        },
        ids: new Set(),
        currentHeading: 1,
        namedExamples: new Map(),
    };

    let currentDoc: Document | null = null;
    const getCurrentDocument = (): Document => {
        if (!currentDoc) {
            throw new Error(
                "Internal error: getting current document before rendering has started",
            );
        }
        return currentDoc;
    };

    const included = new Map<string, string>();
    const md = markdownIt({
        html: true,
    });
    md.use(markdownItDeflist);
    md.use(
        codePreview({
            generateExample: options.generateExample,
        }),
    );
    md.use(headingLevel({ initialLevel: 1 }));
    md.use(
        imageResources({
            addResource: options.addResource,
        }),
    );
    md.use(
        containerRenderer(
            getCurrentDocument,
            docs,
            env,
            included,
            options.handleSoftError,
            {
                messagebox: { title: {}, ...options.messagebox },
            },
        ),
    );
    md.use(table());
    md.use(codeInline());

    return {
        render(doc: DocumentPage, content: string): string {
            currentDoc = doc;
            included.clear();
            included.set(doc.id, doc.id);
            env.fileInfo = doc.fileInfo;
            env.ids = new Set();
            const html = md.render(content, env);
            return processInlineTags(
                inlineTags,
                doc,
                docs,
                html,
                options.handleSoftError,
            );
        },
    };
}
