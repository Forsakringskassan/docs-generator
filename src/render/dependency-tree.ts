import fs from "node:fs/promises";
import { existsSync } from "node:fs";
import path from "node:path";
import { type RenderOptions } from "./render-options";

interface SerializedTreeV1 {
    version: 1;
    tree: Record<string, string[]>;
}

const filename = ".dependency-tree.json";

function replacer(_key: string, value: unknown): unknown {
    if (value instanceof Map) {
        return Object.fromEntries(value.entries());
    }
    if (value instanceof Set) {
        return Array.from(value);
    }
    return value;
}

function isSerializedV1(value: unknown): value is SerializedTreeV1 {
    if (!value || typeof value !== "object") {
        return false;
    }
    return "version" in value && value.version === 1;
}

function parse(value: string): Map<string, Set<string>> {
    try {
        const parsed = JSON.parse(value);
        if (isSerializedV1(parsed)) {
            return new Map(
                Object.entries(parsed.tree).map(([filePath, dependencies]) => {
                    return [filePath, new Set(dependencies)];
                }),
            );
        }
        return new Map();
    } catch {
        return new Map();
    }
}

function serialize(dt: Map<string, Set<string>>): string {
    return JSON.stringify({ version: 1, tree: dt }, replacer, 2);
}

/**
 * Utility class to keep track of build output dependencies.
 *
 * @internal
 */
export class DependencyTree {
    private readonly filePath: string;
    private tree: Map<string, Set<string>>;

    public constructor(options: Pick<RenderOptions, "cacheFolder">) {
        this.filePath = path.join(options.cacheFolder, filename);
        console.log(this.filePath);
        this.tree = new Map();
    }

    public update(filePath: string, dependencies: Set<string>): void {
        this.tree.set(filePath, new Set(dependencies));
    }

    public merge(tree: Map<string, Set<string>>): void {
        for (const [filePath, dependencies] of tree.entries()) {
            this.update(filePath, dependencies);
        }
    }

    public async load(): Promise<void> {
        if (existsSync(this.filePath)) {
            console.log("loading existing dependency tree");
            const content = await fs.readFile(this.filePath, "utf-8");
            this.tree = parse(content);
        } else {
            console.log("skipping loading existing dependency tree");
        }
    }

    public async save(): Promise<void> {
        const content = serialize(this.tree);
        await fs.mkdir(path.dirname(this.filePath), { recursive: true });
        await fs.writeFile(this.filePath, content, "utf-8");
    }

    public getDependenciesFor(
        filePath: string | null,
        options: { recursive?: boolean } = {},
    ): Set<string> {
        const { recursive = false } = options;
        if (!filePath) {
            return new Set();
        }
        const dependencies = new Set(this.tree.get(filePath));
        if (recursive) {
            for (const dependency of dependencies) {
                for (const nested of this.getDependenciesFor(
                    dependency,
                    options,
                )) {
                    dependencies.add(nested);
                }
            }
        }
        return dependencies;
    }
}
