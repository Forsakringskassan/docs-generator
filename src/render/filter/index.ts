export { dump } from "./dump";
export { json } from "./json";
export { relative } from "./relative";
export { take } from "./take";
