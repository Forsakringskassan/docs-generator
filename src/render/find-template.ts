import fs from "node:fs";
import path from "node:path";
import { type DocumentPage, type FileInfo } from "../document";
import { templateDirectory } from "./template-directory";

class MissingTemplateError extends Error {
    private searchDir: string[];

    public constructor(
        fileInfo: FileInfo,
        template: string,
        format: string,
        searchDir: string[],
    ) {
        const message = `Failed to find template "${template}" (${format} format) when rendering "${fileInfo.fullPath}"`;
        super(message);
        this.name = "MissingTemplateError";
        this.searchDir = searchDir;
    }

    public prettyError(): string {
        return [
            this.message,
            "",
            "Searched the following directories:",
            ...this.searchDir.map((it) => `  - ${it}`),
        ].join("\n");
    }
}

const cache = new Map<string, string>();

function cacheKey(layout: string, extension: "html" | "json"): string {
    return [layout, extension].join("|");
}

/**
 * @internal
 * @param from - Which document requested this template (for logging/debugging).
 */
export function findTemplate(
    folders: string[],
    from: FileInfo,
    doc: DocumentPage,
): string;
export function findTemplate(
    folders: string[],
    from: FileInfo,
    layout: string,
    format?: "html" | "json",
): string;
export function findTemplate(
    folders: string[],
    from: FileInfo,
    src: DocumentPage | string,
    format?: "html" | "json",
): string {
    let layout: string;

    if (typeof src === "string") {
        layout = src;
        format ??= "html";
    } else {
        layout = src.template;
        format = src.format !== "json" ? "html" : "json";
    }

    const key = cacheKey(layout, format);
    const cached = cache.get(key);
    if (cached) {
        return cached;
    }

    folders = [...folders, templateDirectory];
    const template = `${layout}.template.${format}`;
    const searchPaths = folders.map((it) => {
        return path.join(it, template);
    });
    const found = searchPaths.find((it) => fs.existsSync(it));
    if (!found) {
        throw new MissingTemplateError(from, template, format, folders);
    }

    cache.set(key, template);
    return template;
}
