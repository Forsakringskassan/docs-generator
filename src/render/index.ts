export { type MarkdownEnv } from "./markdown-env";
export { nunjucksProcessor } from "./nunjucks-processor";
export { type RenderOptions } from "./render-options";
export {
    type SoftErrorCode,
    type SoftErrorDetails,
    type SoftErrorType,
    SoftError,
} from "./soft-error";
