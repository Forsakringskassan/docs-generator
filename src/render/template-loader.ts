import { existsSync } from "node:fs";
import fs from "node:fs/promises";
import path from "node:path";
import { type Callback, type ILoaderAsync, type LoaderSource } from "nunjucks";
import { templateDirectory } from "./template-directory";

interface ResolvedTemplate {
    filePath: string;
    content: string;
}

/**
 * @internal
 */
export class TemplateLoader implements ILoaderAsync {
    public readonly async = true as const;
    private readonly folders: string[];
    private readonly templateCache: Map<string, ResolvedTemplate>;

    public constructor(folders: string[] = []) {
        this.folders = [...folders, templateDirectory];
        this.templateCache = new Map();
    }

    /* eslint-disable-next-line @typescript-eslint/no-misused-promises -- technical debt */
    public async getSource(
        name: string,
        callback: Callback<Error, LoaderSource>,
    ): Promise<void> {
        try {
            const { content, filePath } = await this.resolveTemplate(name);
            callback(null, {
                src: content,
                path: filePath,
                noCache: false,
            });
        } catch (err: unknown) {
            if (err instanceof Error) {
                callback(err, null);
            } else {
                callback(new Error(String(err)), null);
            }
        }
    }

    public hasTemplate(name: string): boolean {
        const { templateCache } = this;
        const cached = templateCache.get(name);
        if (cached) {
            return true;
        }
        const filePath = this.findTemplateFile(name);
        return Boolean(filePath);
    }

    private async resolveTemplate(name: string): Promise<ResolvedTemplate> {
        const { templateCache, folders } = this;
        const cached = templateCache.get(name);
        if (cached) {
            return cached;
        }
        const filePath = this.findTemplateFile(name);
        if (!filePath) {
            const searched = folders.map((it) => `  - "${it}"`).join("\n");
            const message = `Failed to resolve template "${name}", searched in:

${searched}

Make sure the name is correct and the template file exists in one of the listed directories.`;
            throw new Error(message);
        }
        const content = await fs.readFile(filePath, "utf-8");
        const resolved = { content, filePath };
        templateCache.set(name, resolved);
        return resolved;
    }

    private findTemplateFile(name: string): string | undefined {
        const { folders } = this;
        const searchPaths = folders.map((it) => path.join(it, name));
        return searchPaths.find((it) => existsSync(it));
    }
}
