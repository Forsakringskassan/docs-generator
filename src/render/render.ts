import fs from "node:fs/promises";
import { existsSync, copyFileSync } from "node:fs";
import path from "node:path";
import pathPosix from "node:path/posix";
import { promisify } from "node:util";
import { execa } from "execa";
import nunjucks from "nunjucks";
import { type VendorAsset } from "../vendor";
import { type FileInfo, type Document } from "../document";
import {
    type ExampleBatch,
    generateExample,
    ExampleCompileTask,
    ExampleStandaloneTask,
} from "../examples";
import {
    type NavigationLeaf,
    type NavigationSection,
    type NavigationNode,
    isNavigationSection,
    sortNavigationTree,
} from "../navigation";
import { getFingerprint, getOutputFilePath } from "../utils";
import { type RenderOptions } from "./render-options";
import { TemplateLoader } from "./template-loader";
import * as filter from "./filter";
import { findTemplate } from "./find-template";
import { createMarkdownRenderer } from "./create-markdown-renderer";

interface ActiveNavigationLeaf extends NavigationLeaf {
    active: boolean;
}

interface ActiveNavigationSection extends NavigationSection {
    active: boolean;
}

type ActiveNavigationNode = ActiveNavigationLeaf | ActiveNavigationSection;

const scriptPath = path.join(__dirname, "compile-example.js");
let loader: TemplateLoader | null = null;

function haveOutputFile(
    fileInfo: FileInfo,
): fileInfo is FileInfo & { outputName: string } {
    return fileInfo.outputName !== false;
}

function fillActiveNavigation(
    doc: Document,
    node: NavigationSection,
): ActiveNavigationSection;
function fillActiveNavigation(
    doc: Document,
    node: NavigationLeaf,
): ActiveNavigationLeaf;
function fillActiveNavigation(
    doc: Document,
    node: NavigationNode,
): ActiveNavigationNode;
function fillActiveNavigation(
    doc: Document,
    node: NavigationNode,
): ActiveNavigationNode {
    if (isNavigationSection(node)) {
        const children = node.children.map((it) =>
            fillActiveNavigation(doc, it),
        );
        const active = children.some((it) => it.active);
        return {
            ...node,
            active,
            children,
        };
    } else {
        const active = doc.id === node.id;
        return {
            ...node,
            active,
        };
    }
}

function findSidenav(
    doc: Document,
    tree: NavigationSection,
): ActiveNavigationNode | undefined {
    const { path, name } = doc.fileInfo;
    const category = path === "." ? name : path.split("/")[1];
    const sidenav = tree.children
        .filter(isNavigationSection)
        .find((it) => it.key === `./${category}`);
    if (sidenav) {
        return fillActiveNavigation(doc, sidenav);
    }
}

function cache(
    cacheFolder: string,
    outputFolder: string,
): (it: { outputFile: string }) => boolean {
    return (it: { outputFile: string }) => {
        const cacheFile = path.join(cacheFolder, it.outputFile);
        const outputFile = path.join(outputFolder, it.outputFile);
        /* @todo technical debt: shame on me, doing side-effects in a filter function :( and not even async */
        if (existsSync(cacheFile)) {
            copyFileSync(cacheFile, outputFile);
            return false;
        } else {
            return true;
        }
    };
}

async function compileExamples(options: {
    fileInfo: FileInfo;
    cacheFolder: string;
    outputFolder: string;
    tasks: ExampleCompileTask[];
    vendors: VendorAsset[];
}): Promise<void> {
    const { fileInfo, tasks, vendors } = options;
    if (tasks.length === 0) {
        return;
    }

    const cacheFolder = path.posix.join(options.cacheFolder, fileInfo.path);
    const outputFolder = path.posix.join(options.outputFolder, fileInfo.path);
    const cacheMiss = cache(cacheFolder, outputFolder);
    const dirtyTasks = tasks.filter(cacheMiss);

    /* skip forking if there are no tasks to be run anyway */
    if (dirtyTasks.length === 0) {
        return;
    }

    const batch: ExampleBatch = {
        outputFolder,
        external: vendors.map((it) => it.package),
        tasks: dirtyTasks,
    };
    const result = await execa("node", [scriptPath], {
        input: JSON.stringify(batch),
        all: true,
    });
    const hasOutput = result.all ? result.all.length > 0 : false;
    if (hasOutput) {
        /* eslint-disable-next-line no-console -- We don't want to hide how it went  */
        console.log(result.all);
    }
}

async function compileStandalones(options: {
    fileInfo: FileInfo;
    cacheFolder: string;
    outputFolder: string;
    tasks: ExampleStandaloneTask[];
    renderTemplate(name: string, context: object): Promise<string>;
    templateData: Record<string, unknown>;
}): Promise<void> {
    const { fileInfo, tasks, renderTemplate, templateData } = options;
    if (tasks.length === 0) {
        return;
    }

    const cacheFolder = path.posix.join(options.cacheFolder, fileInfo.path);
    const outputFolder = path.posix.join(options.outputFolder, fileInfo.path);
    const cacheMiss = cache(cacheFolder, outputFolder);
    const dirtyTasks = tasks.filter(cacheMiss);

    const standaloneTemplate = findTemplate(fileInfo, "example");
    for (const task of dirtyTasks) {
        const outputFile = path.join(outputFolder, task.outputFile);
        const content = await renderTemplate(standaloneTemplate, {
            ...templateData,
            content: task.content,
        });
        await fs.writeFile(outputFile, content, "utf-8");
    }
}

/**
 * @internal
 */
export function createTemplateLoader(folders: string[]): void {
    loader = new TemplateLoader(folders);
}

/**
 * @internal
 */
export async function render(
    doc: Document,
    docs: Document[],
    nav: { topnav: NavigationSection; sidenav: NavigationSection },
    vendors: VendorAsset[],
    options: RenderOptions,
): Promise<string | null> {
    const { fileInfo } = doc;
    const { outputFolder, cacheFolder } = options;

    /* skip rendering files which have no output */
    if (!haveOutputFile(fileInfo)) {
        return null;
    }

    /* ensure output directory exists */
    const dst = getOutputFilePath(outputFolder, fileInfo);
    const mkdir = fs.mkdir(path.dirname(dst), { recursive: true });

    const topnav = nav.topnav;
    const sidenav = findSidenav(doc, nav.sidenav);
    if (sidenav && isNavigationSection(sidenav)) {
        sortNavigationTree(sidenav);
    }

    const njk = new nunjucks.Environment(loader, { autoescape: false });
    const asyncRender = promisify<string, object, string>(njk.render);
    const renderTemplate = asyncRender.bind(njk);
    const template = findTemplate(doc.fileInfo, doc);
    const templateData = {
        ...options.templateData,
        site: options.site,
        doc,
        topnav,
        rootUrl(doc: Document) {
            const { fileInfo } = doc;
            const relative = pathPosix.relative(fileInfo.path, ".");
            return relative !== "" ? relative : ".";
        },
        sidenav,
        vendors,
    };

    const generatedExamples: ExampleCompileTask[] = [];
    const generatedStandalone: ExampleStandaloneTask[] = [];

    const markdownRenderer = createMarkdownRenderer({
        docs,
        generateExample({ source, language, filename, tags }) {
            const example = generateExample({
                source,
                language,
                filename,
                parent: fileInfo.fullPath,
                setupPath: options.setupPath,
                exampleFolders: options.exampleFolders,
                tags,
            });

            if (example.output) {
                const { dir, name } = path.parse(example.output);
                generatedExamples.push(example.task);
                generatedStandalone.push({
                    outputFile: path.join(dir, `${name}.html`),
                    content: example.markup,
                });
            }

            return example;
        },
        addResource(dst, src) {
            if (!existsSync(src)) {
                throw new Error(`Failed to find image "${src}"`);
            }
            options.addResource(dst, src);
        },
        handleSoftError(error) {
            /* rethrow error so we get hard fails on everything */
            throw error;
        },
    });

    njk.addFilter("marked", (content: string) => {
        return markdownRenderer.render(doc, content);
    });
    njk.addFilter("json", filter.json);
    njk.addFilter("dump", filter.dump);
    njk.addFilter("relative", filter.relative);
    njk.addFilter("take", filter.take);
    njk.addExtension("BlockContainer", {
        tags: ["container"],
        parse(parser, nodes) {
            const tok = parser.nextToken();
            const args = parser.parseSignature(null, true);
            parser.advanceAfterBlockEnd(tok.value);
            return new nodes.CallExtensionAsync(this, "run", args, []);
        },
        async run(
            { ctx }: { ctx: Record<string, unknown> },
            container: string,
            callback: (
                err: Error | null,
                data: nunjucks.runtime.SafeString,
            ) => void,
        ) {
            const blocks = options.templateBlocks.get(container) ?? [];
            const promises = blocks.map(({ renderer }) => {
                if ("filename" in renderer) {
                    const data = renderer.data as object;
                    return renderTemplate(renderer.filename, {
                        ...ctx,
                        ...data,
                    });
                }
                if ("render" in renderer) {
                    return Promise.resolve(renderer.render());
                }
                return Promise.resolve("");
            });
            const content = await Promise.all(promises);
            const markup = content.join("");
            return callback(null, new nunjucks.runtime.SafeString(markup));
        },
    } as nunjucks.Extension);

    await mkdir;

    let content;
    try {
        content = await renderTemplate(template, templateData);
    } catch (err: unknown) {
        const prefix = `Failed to render "${fileInfo.fullPath}"`;
        const message = err instanceof Error ? err.message : String(err);
        throw new Error(`${prefix}: ${message}`, { cause: err });
    }

    const expandedDst = dst.replace(/\[([^]+)\]/g, (match, key) => {
        if (key === "hash") {
            return getFingerprint(doc.body);
        } else {
            return match;
        }
    });
    const writeFile = fs.writeFile(expandedDst, content ?? "null", "utf-8");

    try {
        await compileExamples({
            fileInfo,
            cacheFolder,
            outputFolder,
            tasks: generatedExamples,
            vendors,
        });
    } catch (err: unknown) {
        const prefix = `Failed to compile examples from "${fileInfo.fullPath}"`;
        const message = err instanceof Error ? err.message : String(err);
        throw new Error(`${prefix}: ${message}`, { cause: err });
    }

    try {
        await compileStandalones({
            fileInfo,
            cacheFolder,
            outputFolder,
            tasks: generatedStandalone,
            renderTemplate,
            templateData,
        });
    } catch (err: unknown) {
        const prefix = `Failed to render standalone examples from "${fileInfo.fullPath}"`;
        const message = err instanceof Error ? err.message : String(err);
        throw new Error(`${prefix}: ${message}`, { cause: err });
    }

    await writeFile;
    return dst;
}
