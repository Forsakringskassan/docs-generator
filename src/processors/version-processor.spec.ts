import { promisify } from "node:util";
import nunjucks from "nunjucks";
import { TemplateLoader } from "../render/template-loader";
import { type VersionProcessorTemplateData as TemplateData } from "./version-processor";

const pkg = {
    name: "mock-package",
    version: "1.2.3",
    homepage: "https://example.net",
};

const build = {
    date: "1999-12-31",
    time: "13:37:00",
};

describe("should render correctly", () => {
    const loader = new TemplateLoader(["templates"]);
    const njk = new nunjucks.Environment(loader, { autoescape: false });
    const asyncRender = promisify<string, object, string>(njk.render); // eslint-disable-line @typescript-eslint/unbound-method -- technical debt */
    const renderTemplate = asyncRender.bind(njk);
    const template = "partials/version.njk.html";

    it("without scm data", async () => {
        expect.assertions(1);
        const data: TemplateData = {
            pkg,
            build,
            scm: null,
        };
        const result = await renderTemplate(template, data);
        expect(result).toMatchSnapshot();
    });

    it("without PR data", async () => {
        expect.assertions(1);
        const data: TemplateData = {
            pkg,
            build,
            scm: {
                branch: "feature/foobar",
                commitUrl: "https://example.net/commits/c0ffee",
                commitShort: "c0ffee",
                commitHash: "c0ffeedeadbeeffdfdfdfd",
                pr: undefined,
                prUrl: undefined,
            },
        };
        const result = await renderTemplate(template, data);
        expect(result).toMatchSnapshot();
    });

    it("with all data", async () => {
        expect.assertions(1);
        const data: TemplateData = {
            pkg,
            build,
            scm: {
                branch: "feature/foobar",
                commitUrl: "https://example.net/commits/c0ffee",
                commitShort: "c0ffee",
                commitHash: "c0ffeedeadbeeffdfdfdfd",
                pr: "123",
                prUrl: "https://example.net/pull-requests/123",
            },
        };
        const result = await renderTemplate(template, data);
        expect(result).toMatchSnapshot();
    });
});
