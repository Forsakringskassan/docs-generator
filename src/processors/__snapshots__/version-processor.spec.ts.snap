// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly with all data 1`] = `
"<div class="toolbar__item">
    <span class="version">
        <a href="https://example.net/pull-requests/123">PR 123</a>
        feature/foobar (<a href="https://example.net/commits/c0ffee">c0ffee</a>)<br />
        1999-12-31 13:37:00
    </span>
</div>
"
`;

exports[`should render correctly without PR data 1`] = `
"<div class="toolbar__item">
    <span class="version">
        
        feature/foobar (<a href="https://example.net/commits/c0ffee">c0ffee</a>)<br />
        1999-12-31 13:37:00
    </span>
</div>
"
`;

exports[`should render correctly without scm data 1`] = `
"<div class="toolbar__item">
    <span class="version">
        1.2.3
    </span>
</div>
"
`;
